bleachbit/Cleaner.py

```diff
@@ -479,10 +478,14 @@ def gtk_purge_items():
                     '~/.recently-used.xbel',
                     '~/.local/share/recently-used.xbel*',
                     '~/snap/*/*/.local/share/recently-used.xbel']
-            for path1 in xbel_pathnames:
-                for path2 in glob.iglob(os.path.expanduser(path1)):
-                    if os.path.lexists(path2):
-                        yield Command.Shred(path2)
+            shred_paths = (
```

Comment:
This is tider without having to add more indentation. This syntax is uncommon, though, and I wonder whether other people would have a hard time reading it, so I am not sure this is worth it. What is it called?

Reply:
The syntax here is called a generator expression. It's a "lazy" version of the list comprehension syntax. I think list/generator expressions are relatively popular now-a-days, however I can reset this specific diff hunk with no problems! 👍

Reply:
@kenrap Probably more readable will be if we clean the nested loops with 'map' and 'filter' for example...

Reply:
@rados May I have you provide me an example?

On a technical note, I think Guido would be in disagreement there...

https://www.artima.com/weblogs/viewpost.jsp?thread=98196

Reply:
@kenrap Okay, I see, thank you. Then it could be made flat with two generator expressions. You can give it a try and see if it is more readable...

Reply:
@rados Hmm, with `glob.iglob` returning a glob iterator from each path in `xbel_pathnames`, the problem here is, even if I split it off, I would still need to do that "multi-line generator expression" in order to "flatten" out the glob iterator in the second generator expression. This is because the first generator expression would be a "generator of iterators."

```python
glob_iters = (glob.iglob(os.path.expanduser(path)) for path in xbel_pathnames)
shred_paths = (
    path
    for glob_iter in glob_iters
    for path in glob_iter
    if os.path.lexists(path))
```

I think the above approach would be less desirable?

Reply:
I had an idea of using the REPL to help with visualizing the problem and wanted to share this despite it not being necessary:
```python
>>> import glob, os
>>> paths = ["/var", "/foo", "/mnt"]
>>> glob_iters = [glob.iglob(os.path.expanduser(path)) for path in paths] # pretend this is a generator since they are exhaustible
>>> glob_iters
[<generator object _iglob at 0x7fcb6525b450>, <generator object _iglob at 0x7fcb6525b4c0>, <generator object _iglob at 0x7fcb6525b530>]
>>> flatten = [path for glob_iter in glob_iters for path in glob_iter]
>>> flatten
['/var', '/mnt']
```
Although it's interesting how using `glob.iglob` already filtered out the non-existent path without needing to use `os.path.lexists`.